include/arch/arm/cortex_m/scripts/linker.ld

```diff
@@ -68,6 +68,16 @@
 	#define RAM_ADDR CONFIG_SRAM_BASE_ADDRESS
 #endif
 
+/* Set alignment to CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE
+ * to make linker section alignment comply with MPU granularity.
+ */
+#if defined(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE) \
+	&& (CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0)
+_region_min_align = CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE;
```

Comment:
Instead of checking of the config value is not 0, wouldn't it be better to define it directly as 4 so that the same alignment value is used in both the linker files and the code?

Reply:
For MPUs which require 0-byte alignment, we will still need a a minimum 4-byte alignment both in the linker script and the MPU code. So my point would be to drop the `&& (CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0)` and to make the default value of `CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE` to 4 instead of 0 on non-ARM MPUs. That way both the linker and the MPU code would use the same value.

Reply:
Right. But still someone can mess up and overwrite that with 0. So we would need this extra check, imho.

Reply:
This is actually quite important to have `CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE` != 0 in the MPU code otherwise code like this one would simply not work and truncate the size to 0:

```
_current->stack_info.size &=
		~(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE - 1);
```

Reply:
So, this code block is compiled-in only if MPU support is enabled. So my intention was that, if anyone uses the MPU, and overwrites the definition of CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE, they must know what they are doing. Defining =0 does not make sense (there is no 0-alignment requirement. the minimum could be 1).

Reply:
As you said this is an _hidden_ Kconfig option, so no-one should not be able to mess up. I think it's important to have a consistent value between the MPU code and the linker code. That's actually why it's cleaner to have the value defined in a single place.

Reply:
Though I agree with your argument, I have a single concern: If anyone does not want to use the MPU (builds without MPU support), why would we need to have CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE != 0? I don't see this necessary _in C code_
In the linker, however, we still need alignment-4 for reasons _other than MPU_ That's why the linker enforces 4-byte minimum.

Reply:
OH, I see. Yes, I am convinced that this can be removed. I agree that if anyone sets CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE to zero, deliberately, he should face the consequences of their action.

Reply:
Fixing this @aurel32